src/Sparse/Wiedemann.jl

```diff
@@ -0,0 +1,235 @@
+some_nullspace(A::SMat) = wiedemann(A::SMat, transpose(A)::SMat)
+
+#(p-1)/2 prime 
+@doc Markdown.doc"""
+    wiedemann(A::SMat{gfp_elem}, TA::SMat{gfp_elem}) ->Vector{gfp_elem}
+Computes ker($A$).
+"""
+function wiedemann(A::SMat{gfp_elem}, TA::SMat{gfp_elem}) #N::Int64
+	@info "wiedemann 1 used"
+	RR = base_ring(A)
+	N = modulus(RR)
+	T = elem_type(RR)
+	(n,m) = nrows(A),ncols(A)
+	# Prealloc +Randomchoice
+    r = rand(RR, m)
+	c = rand(RR, m)
+    randlin = rand_srow(min(m,10),m,min(10,N),RR)
+	seq = Vector{T}(undef, 2*n)
+	storing_n = Vector{T}(undef,n)
+    storing_m =  Vector{T}(undef,m)
+	z = zero(RR)
+    #Wiedemann sequence
+    # solve A^tAx = A^tAr = y => x -r in kernel(A^tA) to avoid finding zero vec
+	y = Hecke.mul!(storing_m,TA, Hecke.mul!(storing_n,A,r))
+	seq[1] = dot(randlin,c,z) #randlin*c 		
+	for i = 2:2*n  #Wiedemann sequence 2m?
+        c =  Hecke.mul!(c,TA, Hecke.mul!(storing_n,A,c))
+		seq[i] = dot(randlin,c) 
+	end
+
+	done,f = Hecke.berlekamp_massey(seq)
+	delta =0
+	while iszero(coeff(f,0)) #TODO collect coeffs:
+		delta+=1
+		f = divexact(f,gen(parent(f)))
+	end
+	constpartoff = coeff(f,0)
+	a = -inv(constpartoff)
+	reducedf = divexact(f-constpartoff,gen(parent(f)))
+	#f(TA*A)'c
+    v = Hecke.evaluate(reducedf,TA,A,y).*a
+    return true, (v-r)
+end
+
+@doc Markdown.doc"""
+    wiedemann(A::SMat{gfp_fmpz_elem}, TA::SMat{gfp_fmpz_elem}) ->Vector{gfp_fmpz_elem}
+Computes ker($A$).
+"""
+function wiedemann(A::SMat{gfp_fmpz_elem}, TA::SMat{gfp_fmpz_elem}) #N::fmpz
```

Comment:
Is this not the same function as above?

Reply:
It differs in the computation of the wiedemann sequence. For fmpz I use multi! (defined in the same file below) instead of mul!.

Reply:
OK, thanks for the explanation. For the future, to avoid code duplication, one would do
```
my_mul!(c::Vector{T}, A::SMat{T}, b::Vector{T}) where T <:Union{gfp_fmpz_elem, fmpz_mod} = multi!(c, A, b)

my_mul!(c::Vector{T}, A::SMat{T}, b::Vector{T}) where T <:Union{gfp_elem, nmod} = mul!(c, A, b)
```
and then have only one version of `wiedemann` where `my_mul!` is used.

But this can be done later.